common/src/main/java/com/wynntils/screens/maps/MainMapScreen.java

```diff
@@ -0,0 +1,280 @@
+
+        double distanceX = pX - mapCenterX;
+        double distanceZ = pZ - mapCenterZ;
+
+        float cursorX = (float) (centerX + distanceX * currentZoom);
+        float cursorZ = (float) (centerZ + distanceZ * currentZoom);
+        RenderUtils.MapRenderer.renderCursor(
+                poseStack, cursorX, cursorZ, 1.5f, false, CommonColors.WHITE, PointerType.Arrow);
+    }
+
+    private void renderMap(PoseStack poseStack, MapProfile map, float textureX, float textureZ) {
+        RenderUtils.drawRect(
+                poseStack,
+                CommonColors.BLACK,
+                renderX + renderedBorderXOffset,
+                renderY + renderedBorderYOffset,
+                0,
+                mapWidth,
+                mapHeight);
+        RenderUtils.MapRenderer.renderMapQuad(
+                map,
+                poseStack,
+                centerX,
+                centerZ,
+                textureX,
+                textureZ,
+                mapWidth,
+                mapHeight,
+                1f / currentZoom,
+                false,
+                false);
+    }
+
+    private void updateMapCenterIfDragging(int mouseX, int mouseY) {
+        if (dragging) {
+            float zoomScale = currentZoom;
+            updateMapCenter((float) (mapCenterX + (lastMouseX - mouseX) * zoomScale), (float)
+                    (mapCenterZ + (lastMouseY - mouseY) * zoomScale));
+        }
+        lastMouseX = mouseX;
+        lastMouseY = mouseY;
+    }
+
+    @Override
+    public void renderBackground(PoseStack poseStack) {
+        RenderUtils.drawScalingTexturedRect(
+                poseStack,
+                Texture.FULLSCREEN_MAP_BORDER.resource(),
+                renderX,
+                renderY,
+                100,
+                renderWidth,
+                renderHeight,
+                Texture.FULLSCREEN_MAP_BORDER.width(),
+                Texture.FULLSCREEN_MAP_BORDER.height());
+    }
+
+    @Override
+    public boolean mouseScrolled(double mouseX, double mouseY, double delta) {
+        setZoom((float) (currentZoom + delta * ZOOM_FACTOR));
+
+        return true;
+    }
+
+    @Override
+    public boolean mouseClicked(double mouseX, double mouseY, int button) {
+        if (button == GLFW.GLFW_MOUSE_BUTTON_RIGHT) {
+            centerMapAroundPlayer();
+        } else if (button == GLFW.GLFW_MOUSE_BUTTON_LEFT) {
+            dragging = true;
+            lastMouseX = mouseX;
+            lastMouseY = mouseY;
+        } else if (button == GLFW.GLFW_MOUSE_BUTTON_MIDDLE) {
+            setCompassToMouseCoords(mouseX, mouseY);
+        }
+
+        return true;
+    }
+
+    private void setCompassToMouseCoords(double mouseX, double mouseY) {
+        double gameX = (mouseX - centerX) / currentZoom + mapCenterX;
+        double gameZ = (mouseY - centerZ) / currentZoom + mapCenterZ;
+        Location compassLocation = new Location(gameX, 0, gameZ);
+        CompassModel.setCompassLocation(compassLocation);
```

Comment:
Bonus points for this! :) It's a clear sign that we've done stuff right when it is so easy to implement.